<a href="https://colab.research.google.com/github/notKguin/MMK-Informservice/blob/main/final_project/project/ML_intensive_final_project_cleaned.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Настройки

In [1]:
# pd.set_option('display.max_columns', None) # Убираем ограничение в количестве столбцов
# pd.set_option('display.max_rows', None) # Убираем ограничение в количестве строк

In [2]:
# pd.reset_option('display.max_columns') # Сбрасываем ограничение по умолчанию в количестве столбцов
# pd.reset_option('display.max_rows') # Сбрасываем ограничение по умолчанию в количестве строк

# Загрузка датасета

In [3]:
from google.colab import drive
import pandas as pd
drive.mount('/content/drive')

Mounted at /content/drive


In [4]:
df = pd.read_csv('/content/drive/MyDrive/MMK_inforservice/datasets/Dataset_DP10_day.csv', sep=';', encoding='cp1251', index_col='Unnamed: 0')

In [5]:
df

Time Iron_count          URK  Расход холодного дутья ( м?/час )  \
1     02.09.2012   4359,847   457,062565                        214677.1855   
2     03.09.2012   4425,385  450,1093358                        214572.0651   
3     04.09.2012   4372,657  464,1847051                        216907.2760   
4     05.09.2012   4447,549   455,028522                        217341.6706   
5     06.09.2012   4441,511  467,2329351                        223169.5814   
...          ...        ...          ...                                ...   
1850  26.09.2017   5141,843  404,5660885                        264492.2077   
1851  27.09.2017   5251,577  401,6168941                        268069.9850   
1852  28.09.2017   5305,734  401,4540218                        267186.0690   
1853  29.09.2017   5360,082   402,017341                        266981.6263   
1854  30.09.2017   5288,666  407,7516705                        269118.9583   

     Давление горячего дутья ( ат ) Температура горячего дутья ( С )  \
1                             4(ат)                   1055,686386(С)   
2                             4(ат)                   1063,769384(С)   
3                             4(ат)                   1060,716365(С)   
4                             4(ат)                   1058,044291(С)   
5                             4(ат)                   1060,088181(С)   
...                             ...                              ...   
1850                          4(ат)                   1002,055957(С)   
1851                          4(ат)                   1002,480761(С)   
1852                          4(ат)                   1005,178185(С)   
1853                          4(ат)                   1002,133046(С)   
1854                          4(ат)                   1002,953089(С)   

     Расход природного газа ( м?/час ) Температура природного газа ( ?С )  \
1                          15916,95805                        14,13750247   
2                          15998,28292                        14,47928015   
3                          15984,38981                         14,4863259   
4                          15975,46354                        15,82185074   
5                          15999,31665                        18,93565452   
...                                ...                                ...   
1850                       29951,71387                        4,641558896   
1851                       29905,31812                        6,578101267   
1852                       29838,85042                        5,593210797   
1853                       29622,66935                         5,35199336   
1854                       29912,03947                        4,774703225   

     Содержание кислорода в дутье ( % ) Расход пара на увлажнение ( т/час )  \
1                           26,30299997                         0,145087693   
2                           26,50044982                         0,131810737   
3                           26,51146475                         0,117763162   
4                           26,62238797                         0,128866268   
5                            26,3542157                         0,129628953   
...                                 ...                                 ...   
1850                        28,24363685                         0,274633718   
1851                        28,29148412                                 NaN   
1852                        28,14871367                                 NaN   
1853                        27,90994938                                 NaN   
1854                        27,71687436                                 NaN   

      ...         TiO2          Osn            P           Zn           Cr  \
1     ...  0,273174603  1,885476182  0,027623016  0,031146826  0,023928571   
2     ...  0,253095238  1,870238091  0,026238096   0,03047619  0,022357143   
3     ...  0,254659441  1,898095966  0,026314242  0,029332818   0,02200774   
4     ...  0

# Описание датасета

---
Период исследования: 02.09.2012 - 30.09.2017

---

96 столбцов | 1676 строк  
Индексы с 1 по 1854

---


Описание столбцов:  
1. Дата - пустых значений: 0, формат: object.
2. Произведено чугуна - пустых значений: 0, формат: object.
3. Удельный расход кокса - пустых значений: 0, формат: object.
4. Расход холодного дутья (кубических метров в час) - пустых значений: 0, формат: float64.
5. Давление (атмосфер) - пустых значений: 0, формат: int.
6. Температура горячего дутья (градусов Цельсия) - пустых значений: 0, формат: object.
7. Расход природного газа (кубических метров в час) - пустых значений: 0, формат: object.
8. Температура природного газа (градусов Цельсия) - пустых значений: 0, формат: object.
9. Содержание кислорода в дутье (процентов) - пустых значений: 177, формат: object.
10. Расход пара на увлажнение (тонн в час) - пустых значений: 252, формат: object.
11. Перепад давлений общий (атмосфер) - пустых значений: 0, формат: object.
12. Температура чугуна на выпуске (лётка 1) (градусов Цельсия) - пустых значений: 298, формат: object.
13. Температура чугуна на выпуске (лётка 2) (градусов Цельсия) - пустых значений: 395, формат: object.
14. Давление колошникового газа (атмосфер) - пустых значений: 187, формат: object.
15. Перепад давлений нижний (атмосфер) - пустых значений: 0, формат: object.
16. Перепад давлений верхний (атмосфер) - пустых значений: 0, формат: object.
17. Выход колошникового газа (кубических метров в час) - пустых значений: 187, формат: object.
18. Положение клапана малый снорт (процентов) - пустых значений: 101, формат: object.
19. Температура колошникового газа 1 (градусов Цельсия) - пустых значений: 0, формат: object.
20. Температура колошникового газа 2 (градусов Цельсия) - пустых значений: 0, формат: object.
21. Температура колошникового газа 3 (градусов Цельсия) - пустых значений: 7, формат: object.
22. Температура колошникового газа 4 (градусов Цельсия) - пустых значений: 0, формат: object.
23. Температура колошникового газа в пылеуловителе (градусов Цельсия), формат: object.
24. Уровень засыпи (запад) (метров) - пустых значений: 81, формат: object.
25. Уровень засыпи (восток) (метров) - пустых значений: 0, формат: object.
26. Содержание CO в колошниковом газе (процентов) - пустых значений: 156, формат: object.
27. Содержание CO2 в колошниковом газе (процентов) - пустых значений: 307, формат: object.
28. Содержание H2 в колошниковом газе (процентов) - пустых значений: 115, формат: object.
29. Температура периферийных газов 1 (градусов Цельсия) - пустых значений: 220, формат: object.
30. Температура периферийных газов 2 (градусов Цельсия) - пустых значений: 178, формат: object.
31. Температура периферийных газов 3 (градусов Цельсия) - пустых значений: 226, формат: object.
32. Температура периферийных газов 4 (градусов Цельсия) - пустых значений: 225, формат: object.
33. Температура периферийных газов 5 (градусов Цельсия) - пустых значений: 211, формат: object.
34. Температура периферийных газов 6 (градусов Цельсия) - пустых значений: 244, формат: object.
35. Уровень засыпи (север) (метров) - пустых значений: 229, формат: object.
36. Уровень засыпи (юг) (метров) - пустых значений: 210, формат: object.
37. Содержание N2 в колошниковом газе (процентов) - пустых значений: 54, формат: object.
38. Вода на газоочистку.Расход (кубических метров в час) - пустых значений: 1202, формат: int.
39. Вода на охлаждение.Водовод 1.Расход (кубических метров в час) - пустых значений: 1202, формат: float64.
40. Вода на охлаждение.Водовод 2.Расход (кубических метров в час) - пустых значений: 1202, формат: object.
41. Вода на охлаждение.Водовод 3.Расход ( м?/час ) - пустых значений: 1202, формат: object.
42. Длительность - пустых значений: 0, формат: object.
43. Налив - пустых значений: 456, формат: object.
44. Шлак - пустых значений: 456, формат: object.
45. Чугун.Cr - пустых значений: 456, формат: object.
46. Чугун.Cu - пустых значений: 675, формат: object.
47. Чугун.Mn - пустых значений: 456, формат: object.
48. Чугун.Ni - пустых значений: 675, формат: object.
49. Чугун.P - пустых значений: 456, формат: object.
50. Чугун.S - пустых значений: 456, формат: object.
51. Чугун.Si - пустых значений: 456, формат: object.
52. Чугун.Ti - пустых значений: 456, формат: object.
53. Чугун.V - пустых значений: 456, формат: object.
54. Чугун.C - пустых значений: 470, формат: object.
55. Шлак.Al2O3 - пустых значений: 459, формат: object.
56. Шлак.CaO - пустых значений: 459, формат: object.
57. Шлак.FeO - пустых значений: 459, формат: object.
58. Шлак.Ko - пустых значений: 459, формат: object.
59. Шлак.MgO - пустых значений: 459, формат: object.
60. Шлак.MnO - пустых значений: 459, формат: object.
61. Шлак.NaO - пустых значений: 459, формат: object.
62. Шлак.S - пустых значений: 459, формат: object.
63. Шлак.SiO2 - пустых значений: 459, формат: object.
64. Шлак.TiO2 - пустых значений: 459, формат: object.
65. updp - пустых значений: 456, формат: object.
66. Бар М10 - пустых значений: 0, формат: object.
67. Бар М25 - пустых значений: 0, формат: object.
68. Влага - пустых значений: 0, формат: object.
69. Зола - пустых значений: 0, формат: object.
70. кокс +80 - пустых значений: 0, формат: object.
71. кокс -25 - пустых значений: 0, формат: object.
72. кокс 40-25 - пустых значений: 0, формат: object.
73. кокс 60-40 - пустых значений: 0, формат: object.
74. кокс 80-60 - пустых значений: 0, формат: object.
75. Летучие - пустых значений: 0, формат: object.
76. Сера К - пустых значений: 2, формат: object.
77. Углерод - пустых значений: 2, формат: object.
78. CRI - пустых значений: 219, формат: object.
79. CSR - пустых значений: 219, формат: object.
80. Fe - пустых значений: 0, формат: object.
81. FeO - пустых значений: 0, формат: object.
82. S - пустых значений: 0, формат: object.
83. SiO2 - пустых значений: 0, формат: object.
84. CaO - пустых значений: 0, формат: object.
85. MgO - пустых значений: 0, формат: object.
86. Al2O3 - пустых значений: 0, формат: object.
87. TiO2 - пустых значений: 0, формат: object.
88. Osn - пустых значений: 0, формат: object.
89. P - пустых значений: 0, формат: object.
90. Zn - пустых значений: 0, формат: object.
91. Cr - пустых значений: 0, формат: object.
92. MnO - пустых значений: 0, формат: object.
93. O2 Горячее дутьё - пустых значений: 86, формат: object.
94. Доля агломерата - пустых значений: 0, формат: object.
95. Колошниковый газ - пустых значений: 0, формат: object.
96. Колошниковая пыль - пустых значений: 0, формат: object.

---

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1676 entries, 1 to 1854
Data columns (total 96 columns):
 #   Column                                                 Non-Null Count  Dtype  
---  ------                                                 --------------  -----  
 0   Time                                                   1676 non-null   object 
 1   Iron_count                                             1676 non-null   object 
 2   URK                                                    1676 non-null   object 
 3   Расход холодного дутья ( м?/час )                      1676 non-null   float64
 4   Давление горячего дутья ( ат )                         1676 non-null   object 
 5   Температура горячего дутья ( С )                       1676 non-null   object 
 6   Расход природного газа ( м?/час )                      1676 non-null   object 
 7   Температура природного газа ( ?С )                     1676 non-null   object 
 8   Содержание кислорода в дутье ( % )                   

# Предобработка

In [7]:
df = df.reset_index(drop=True) # Сбрасываем индекс т.к. их больше чем количество строк (из-за пропуска в датах)

In [8]:
df.duplicated().sum() # Проверяем на дубликаты

np.int64(0)

## Обрабатываем столбцы

### Меняем названия

In [9]:
df = df.rename(columns={
    'Time': 'Дата',
    'Iron_count': 'Количество чугуна (т)',
    'URK': 'Удельный расход кокса (кг/т)',
    'Расход холодного дутья ( м?/час )': 'Расход холодного дутья (м³/ч)',
    'Давление горячего дутья ( ат )': 'Давление горячего дутья (ат)',
    'Температура горячего дутья ( С )': 't горячего дутья (°С)',
    'Расход природного газа ( м?/час )': 'Расход природного газа (м³/ч)',
    'Температура природного газа ( ?С )': 't природного газа (°С)',
    'Содержание кислорода в дутье ( % )': 'Содержание кислорода в дутье (%)',
    'Расход пара на увлажнение ( т/час )': 'Расход пара на увлажнение (т/ч)',
    'Перепад давлений общий ( ат )': 'Перепад давлений общий (ат)',
    'Температура чугуна на выпуске (лётка 1) ( ?С )': 't чугуна на выпуске (лётка 1) (°С)',
    'Температура чугуна на выпуске (лётка 2) ( ?С )': 't чугуна на выпуске (лётка 2) (°С)',
    'Давление колошникового газа ( ат )': 'Давление колошникового газа (ат)',
    'Перепад давлений нижний ( ат )': 'Перепад давлений нижний (ат)',
    'Перепад давлений верхний ( ат )': 'Перепад давлений верхний (ат)',
    'Выход колошникового газа ( м?/час )': 'Выход колошникового газа (м³/ч)',
    'Положение клапана малый снорт ( % )': 'Положение клапана малый снорт (%)',
    'Температура колошникового газа 1 ( ?С )': 't колошникового газа 1 (°С)',
    'Температура колошникового газа 2 ( ?С )': 't колошникового газа 2 (°С)',
    'Температура колошникового газа 3 ( ?С )': 't колошникового газа 3 (°С)',
    'Температура колошникового газа 4 ( ?С )': 't колошникового газа 4 (°С)',
    'Температура колошникового газа в пылеуловителе ( ?С )': 't колошникового газа в пылеуловителе (°С)',
    'Уровень засыпи (запад) ( м )': 'Уровень засыпи (запад) (м)',
    'Уровень засыпи (восток) ( м )': 'Уровень засыпи (восток) (м)',
    'Содержание CO в колошниковом газе ( % )': 'CO в колошниковом газе (%)',
    'Содержание CO2 в колошниковом газе ( % )': 'CO₂ в колошниковом газе (%)',
    'Содержание H2 в колошниковом газе ( % )': 'H₂ в колошниковом газе (%)',
    'Температура периферийных газов 1 ( ?С )': 't периферийных газов 1 (°С)',
    'Температура периферийных газов 2 ( ?С )': 't периферийных газов 2 (°С)',
    'Температура периферийных газов 3 ( ?С )': 't периферийных газов 3 (°С)',
    'Температура периферийных газов 4 ( ?С )': 't периферийных газов 4 (°С)',
    'Температура периферийных газов 5 ( ?С )': 't периферийных газов 5 (°С)',
    'Температура периферийных газов 6 ( ?С )': 't периферийных газов 6 (°С)',
    'Уровень засыпи (север) ( м )': 'Уровень засыпи (север) (м)',
    'Уровень засыпи (юг) ( м )': 'Уровень засыпи (юг) (м)',
    'Содержание N2 в колошниковом газе ( % )': 'N₂ в колошниковом газе (%)',
    'Вода на охлаждение.Водовод 1.Расход ( м?/час )': 'Вода на охлаждение. Водовод 1. Расход (м³/ч)',
    'Вода на охлаждение.Водовод 2.Расход ( м?/час )': 'Вода на охлаждение. Водовод 2. Расход (м³/ч)',
    'Вода на охлаждение.Водовод 3.Расход ( м?/час )': 'Вода на охлаждение. Водовод 3. Расход (м³/ч)',
    'dlitelnost': 'Длительность простоя (сут)',
    'Shlak': 'Шлак',
    'Шлак.Al2O3': 'Шлак.Al₂O₃',
    'Шлак.SiO2': 'Шлак.SiO₂',
    'Шлак.TiO2': 'Шлак.TiO₂',
    'updp': 'упдп',
    'кокс +80': 'Кокс +80 мм',
    'кокс -25': 'Кокс -25 мм',
    'кокс 40-25': 'Кокс 40-25 мм',
    'кокс 60-40': 'Кокс 60-40 мм',
    'кокс 80-60': 'Кокс 80-60 мм',
    'SiO2': 'SiO₂',
    'Al2O3': 'Al₂O₃',
    'TiO2': 'TiO₂',
    'O2 Горячее дутьё': 'O₂ в горячем дутье',
    'Доля.агломерата': 'Доля агломерата',
})

### Меняем тип данных

In [10]:
df['Дата'] = pd.to_datetime(df['Дата'], format='%d.%m.%Y')
df['Давление горячего дутья (ат)'] = df['Давление горячего дутья (ат)'].str.replace('\(ат\)', '', regex=True).astype(int)
df['t горячего дутья (°С)'] = df['t горячего дутья (°С)'].str.replace(r'[^\d,]', '', regex=True).str.replace(',', '.').astype(float)

In [11]:
float_cols = [col for col in df.columns if col not in ['Дата', 'Давление горячего дутья (ат)', 'Колошниковый газ', 'Колошниковая пыль']]
df[float_cols] = df[float_cols].replace(',', '.', regex=True).astype(float)

In [12]:
import json
import numpy as np
import re

In [13]:
def parse_gas(row):
    try:
        cleaned_row = row
        for key in ['CO2', 'CO', 'H2']:
            cleaned_row = cleaned_row.replace(f'"{key}": ', f'"{key}":')
            cleaned_row = re.sub(r'(\d+),(\d+)', r'\1.\2', cleaned_row)

        gas_dict = json.loads(cleaned_row)

        return pd.Series({
            'CO2': float(gas_dict['CO2']) if gas_dict['CO2'] is not None else np.nan,
            'CO': float(gas_dict['CO']) if gas_dict['CO'] is not None else np.nan,
            'H2': float(gas_dict['H2']) if gas_dict['H2'] is not None else np.nan
        })
    except Exception as e:
        return pd.Series({'CO2': np.nan, 'CO': np.nan, 'H2': np.nan})

In [14]:
df[['Колошниковый газ (CO₂)', 'Колошниковый газ (CO)', 'Колошниковый газ (H₂)']] = df['Колошниковый газ'].apply(parse_gas)

### Удаляем ненужные столбцы

In [15]:
df = df.drop('Naliv', axis=1) # Подсказка от компании что он не нужен

In [16]:
df = df.drop('Вода на газоочистку.Расход ( м?/час )', axis=1) # Аномальные значения (9999)

In [17]:
df = df.drop('Колошниковый газ', axis=1) # Столбец уже обработан

In [18]:
df = df.drop('Колошниковая пыль', axis=1) # Слишком мало столбцов со значениями (167)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1676 entries, 0 to 1675
Data columns (total 95 columns):
 #   Column                                        Non-Null Count  Dtype         
---  ------                                        --------------  -----         
 0   Дата                                          1676 non-null   datetime64[ns]
 1   Количество чугуна (т)                         1676 non-null   float64       
 2   Удельный расход кокса (кг/т)                  1676 non-null   float64       
 3   Расход холодного дутья (м³/ч)                 1676 non-null   float64       
 4   Давление горячего дутья (ат)                  1676 non-null   int64         
 5   t горячего дутья (°С)                         1676 non-null   float64       
 6   Расход природного газа (м³/ч)                 1676 non-null   float64       
 7   t природного газа (°С)                        1676 non-null   float64       
 8   Содержание кислорода в дутье (%)              1499 non-null   float6

In [20]:
df

Дата  Количество чугуна (т)  Удельный расход кокса (кг/т)  \
0    2012-09-02               4359.847                    457.062565   
1    2012-09-03               4425.385                    450.109336   
2    2012-09-04               4372.657                    464.184705   
3    2012-09-05               4447.549                    455.028522   
4    2012-09-06               4441.511                    467.232935   
...         ...                    ...                           ...   
1671 2017-09-26               5141.843                    404.566088   
1672 2017-09-27               5251.577                    401.616894   
1673 2017-09-28               5305.734                    401.454022   
1674 2017-09-29               5360.082                    402.017341   
1675 2017-09-30               5288.666                    407.751670   

      Расход холодного дутья (м³/ч)  Давление горячего дутья (ат)  \
0                       214677.1855                             4   
1                       214572.0651                             4   
2                       216907.2760                             4   
3                       217341.6706                             4   
4                       223169.5814                             4   
...                             ...                           ...   
1671                    264492.2077                             4   
1672                    268069.9850                             4   
1673                    267186.0690                             4   
1674                    266981.6263                             4   
1675                    269118.9583                             4   

      t горячего дутья (°С)  Расход природного газа (м³/ч)  \
0               1055.686386                    15916.95805   
1               1063.769384                    15998.28292   
2               1060.716365                    15984.38981   
3               1058.044291                    15975.46354   
4               1060.088181                    15999.31665   
...                     ...                            ...   
1671            1002.055957                    29951.71387   
1672            1002.480761                    29905.31812   
1673            1005.178185                    29838.85042   
1674            1002.133046                    29622.66935   
1675            1002.953089                    29912.03947   

      t природного газа (°С)  Содержание кислорода в дутье (%)  \
0                  14.137502                         26.303000   
1                  14.479280                         26.500450   
2                  14.486326                         26.511465   
3                  15.821851                         26.622388   
4                  18.935655                         26.354216   
...                      ...                               ...   
1671                4.641559                         28.243637   
1672                6.578101                         28.291484   
1673                5.593211                         28.148714   
1674                5.351993                         27.909949   
1675                4.774703                         27.716874   

      Расход пара на увлажнение (т/ч)  ...       Osn         P        Zn  \
0                            0.145088  ...  1.885476  0.027623  0.031147   
1                            0.131811  ...  1.870238  0.026238  0.030476   
2                            0.117763  ...  1.898096  0.026314  0.029333   
3                            0.128866  ...  1.900649  0.026509  0.030535   
4                            0.129629  ...  1.911628  0.024758  0.030446   
...                               ...  ...       ...       ...       ...   
1671                         0.274634  ...  1.822452  0.032123  0.040913   
1672                              NaN  ...  1.850458  0.032264  0.042306   
1673                              NaN  ...  1.808062  0.033479  0.041188   
1674               

# Скачиваем датасет

In [21]:
from google.colab import files

In [22]:
df.to_csv('ML-intensive_final_project_cleaned.csv', index=False)

In [23]:
files.download('ML-intensive_final_project_cleaned.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>